swarm/utils/metrics.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+//NOTE: The metrics package for swarm should align with the general go-ethereum metrics,
+//which uses the `rcrowley/go-metrics` package. The latter has limitations with
+//aggregating and resetting metrics, which sparked some research on the swarm team
+//regarding metrics. Therefore, this swarm metrics package introduces some indirection,
+//which abstracts a metrics API, allowing to use a different metrics library if needed.
+//If ultimately the go-metrics issues are solved,
+//it may be omitted in favor of directly using the `metrics/metrics` package in go-ethereum
+
+package utils
+
+import (
+	"github.com/ethereum/go-ethereum/metrics"
+	"net"
+	"time"
+
+	graphite "github.com/cyberdelia/go-metrics-graphite"
+	gometrics "github.com/rcrowley/go-metrics"
+)
+
+var (
+	MetricsEnabled = metrics.Enabled
+	bzzAccount     string
+	SwarmRegistry  gometrics.Registry
+)
+
+type MetricsTimer struct {
+	timer  gometrics.Timer
+	bucket string
```

Comment:
I am curious why is bucket in MetricsTimer? As far as I can see it is only checked for empty string in SendTimer.

Reply:
That is correct. `SendTimer` is used for sending a `timer-end` value (an end to a time measurement), so when creating the timer we store the bucket  for later reuse)

Reply:
Maybe currently is not needed, but if you plan to use it in the future, it is fine for me.